### ISSUE TITLE
Add support for home document and profile

### ISSUE THREAD

**Issue by hibaymj** [katharsis-project/katharsis-framework#322]
Katharsis' stated goal is to add HATEOAS support, as such we should further embrace this goal by adding the required documentation and discovery support to allow the framework to progress down this path.

Adding these should allow us to create a UI for the service in a much more concise and simpler fashion.

For now I propose the use of the [JSON Home document](https://tools.ietf.org/html/draft-nottingham-json-home-02) along with [ALPS](https://tools.ietf.org/html/draft-amundsen-richardson-foster-alps-01) for profiling.

**Comment by chb0github** [comment]
I have long called for this.

**Comment by hibaymj** [comment]
I think we should also make sure we interface the modules, so support for additional or multiple formats of either concern could be supported.

**Comment by chb0github** [comment]
The first thing that needs doing is options support. There is at least one
ticket for that already.

The next thing is this root document. The sample code already shows you how
to get all repositories. The problem is that there is no discoverable
relationship between a root resource '/' and a child resource likes
'/tasks' how would Katharsis know it's a one-to-many? Currently, that's
application specific. A mechanism for describing it must be used.
Alternatively you can take a  one size fits all approach and just assume
everything is a many-to-many that sometimes only ever relates to one other
resource.

I think a good start would be a prototype based on our current examples.

Michael , are you game for taking a stab at this? You obviously have a good
grasp of this and a real desire.

**Comment by hibaymj** [comment]
#47 has been sitting out there for a while.  And while the CORS issue is a whole other thing (which is the only time I've actively seen options used), I don't see how it would impede the consumption of the rest of the service, and contextually HEAD is a bit ambiguous as far as JsonApi spec is concerned.  However, I don't see how this issue is a blocker in this case.

I have already implemented this once, albeit in a proprietary project, so I could probably shortcut some of the dev time with that experience.  However, I'm not sure what you are referring to when you say no discoverable relationship between / and /****.  The reason for suggesting the format is that the home document is not serialized in JsonApi format at all.  The suggestion is to use the Json Home mediatype because that format does support this scenario, where JsonApi would require some form of hack or (undefined as yet) extension.

**Comment by chb0github** [comment]
You need OPTIONS support so you know what verbs are supported on a
resource. CORS uses it for the same reason. You need to know if you can
support POST before clicking submit.

How about a prototype? This is academic without code. And this debate about
UI support has been going on forever

**Comment by hibaymj** [comment]
I guess we could quibble over the 'need' in your sentence, but because GET can function the same way for discovery I didn't evaluate it as a dependency.

Prototype requires being at home to work on it, sadly daytime contributions are reserved to GH while I have an external blocker or build running.  I'll see what I can do, I've already assigned this issue to myself as I presumed it would fall on me to implement since I was the driving requester.

**Comment by remmeier** [comment]
the support for @hibaymj is a good point. 

I think to modules should be setup. A HomeModule in a katharsis-home project and maybe AlpsModule in a katharsis-alps project. The later would hook into the former.

Regarding a katharsis-alps, katharsis-meta provides the same feature set (exposes all the repositories, attributes, types, etc.). A benefit of katharsis-meta is that it is pure JSON-API, making it easy to consume for clients already using JSON API anyway. The drawback is that there is no standard in this area like ALPS is. katharsis-meta could hook into katharsis-home the same way as a potential katharsis-alps would.

**Comment by hibaymj** [comment]
I will certainly be spending some more time deep diving into the meta module to see what I can reuse to allow for easier development.

Regarding the modules, I think individual modules for each standard might be a little bit overkill.  These modules shouldn't be large when done, so it is possible a single module could hold them all for prototyping purposes.

**Comment by remmeier** [comment]
for prototyping sure, but at the end home and alps must to be to separate modules as they are two separate concerns. People may choose to use the former without the later, or replace the later with something different. Changes/extensions to katharsis-core should be limited as well, as it is already big enough. Likely a small thing will be necessary to make it possible to return the home document upon a request.

**Comment by remmeier** [comment]
to some degree I think a katharsis-alps could build on top of katharsis-meta. katharsis-jpa does this as well. It provides a convience layer on top of the katharsis internals. And the type model is more flexible, something where the current ResourceInformation/ResourceField need to catch up at some point in the future. For katharsis-home it would be possible as well, but since it is rather trivial implement it is likely not worth the dependency. Using katharsis-meta not necessary means integrating the module and exposing everythign as json api. It can also be used just interally/programmatically.

**Comment by hibaymj** [comment]
Point granted.

Perhaps we could say 2 modules when finished, katharsis-home, and katharsis-profile.  Home could subsume or rename the meta module (with refactoring) as they are different implementations towards the same goal.  While profile would leave us the opportunity to declare profiles in ALPS or even schema.org and json-ld format.

**Comment by remmeier** [comment]
I think home and meta are mostly separate.

- home provides a root page under the typical "/api"
- while meta is a katharsis-specific alternative to alps

**Comment by hibaymj** [comment]
Ok, I think I misunderstood the goal of the meta module as described in your previous posts.  In that case, the association was wrong but the concept remains.  Home module could take ownership of all ways to produce a home document, now and in the future, while profile could include all the ways to advertise the vocabulary.

**Comment by remmeier** [comment]
exactly, the home module should be fairly easy to setup, there already is an example in the spring boot example I think

**Comment by hibaymj** [comment]
I hope to be able to begin working on this tomorrow evening at the earliest.  I've tentatively added a milestone of 3.1.0 to account for limited development time.

**Comment by chb0github** [comment]
We don't need another sub-module though consisting of a handful of files like the validation module. Root document support should be available by default as should input validation.
